src/share/poudriere/common.sh

```diff
@@ -6877,6 +6881,64 @@ load_priorities() {
 	return 0
 }
 
+pkg_seeding() {
+	[ "${PKG_SEEDING}" != "no" ] || return 0
+
+	[ "${PWD}" = "${MASTERMNT}/.p" ] || \
+	    err 1 "pkg_seeding requires PWD=${MASTERMNT}/.p"
+
+	local url url_prefix jail_version jail_arch dest
+	local seed_total seed_ok
+
+	# Set up URL prefix
+	jail_version=$(jget ${JAILNAME} version)
+	# $MARJOR_RELEASE-$SUBRELEASE.RELEASE-px --> $MARJOR_RELEASE
+	jail_version="${jail_version%%.*}"
+	jail_arch=$(jget ${JAILNAME} arch)
+	url_prefix="${PKG_SEEDING_METHOD}://${PKG_SEEDING_MIRROR}"
+	url_prefix="${url_prefix}/FreeBSD:${jail_version}:${jail_arch}/latest/All"
+
+	msg "Package seeding from ${PKG_SEEDING_MIRROR}"
+	bset status "packageseeding:"
+
+	seed_ok=0
+	parallel_start
+	while read pkgname originspec dep_reason; do
+		# Don't fetch ports explicitly requested to be built
+		# Only fetch dependencies the users don't bother being built
+		# by others
+		pkgname="${pkgname}.${PKG_EXT}"
+		url="${url_prefix}/${pkgname}"
+		dest="${PACKAGES}/All/${pkgname}"
+		if [ "${dep_reason}" != "listed" ] && [ ! -f "$dest" ]; then
+			parallel_run pkg_seeding_download "${url}" "${dest}" "${originspec}" && \
+                seed_ok=$((${seed_ok} + 1))
+		fi
+	done < "all_pkgs"
+
+	if ! parallel_stop; then
+		err 1 "Fatal errors encountered seeding packages."
+	fi
+
+	msg "Package seeding done, seeded ${seed_ok} packages."
+}
+
+pkg_seeding_download() {
+	[ $# -ne 3 ] && eargs pkg_seeding_download destination url originspec
+
+	msg_debug "Fetching seed from: $1"
+	fetch -q -a -w 1 -m -o "$2" "$1" &> /dev/null || \
```

Comment:
this could do with a comment; what value does the 2nd retry with different parameters bring here? does it solve an actual problem, and in the event of a failure can we rely on the retrieved package being correct? Should we check anywhere for validity, whether by SHA* or other method?

Reply:
This kind of fetch retry is actually an imitation of the `fetch_file` function defined in the same file.  Its purpose is to recover from potential jitterish network that fails the first fetch.  However, we cannot verify the validity of fetched package without checking its SHA against pkg.freebsd.org's `digests.txz`.  As this will add quite a bit complexity, I will backlog it for now and finish the most basic functionality first. 😸